net/isoping/isoping_test.go

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) 2021 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+package isoping
+
+import (
+	"bytes"
+	"encoding/binary"
+	"math"
+	"testing"
+)
+
+// Tests if our stddev calculation is within reason
+// Must do some rounding to a certain significant digit
+// Currently only need 6 digits for the testing.
+func sigDigs(x float64, digs int) float64 {
+	return math.Round(x*math.Pow10(digs)) / math.Pow10(digs)
+}
+
+// TestOnepass_stddev tests if the function receives the same answer as in
+// the C implementation of this function.
+func TestOnepass_stddev(t *testing.T) {
+	t.Parallel()
+	tests := []struct {
+		name  string
+		input []int64
+		out   float64
+	}{
+
+		{
+			name:  "basic1",
+			input: []int64{12, 2, 3},
+			out:   2.309401,
+		},
+		{
+
+			name:  "basic2",
+			input: []int64{12023232232, 212, 321},
+			out:   6129.649279,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			ttAns := sigDigs(onePassStddev(tt.input[0], tt.input[1], tt.input[2]), 6)
+			if ttAns != tt.out {
+				t.Errorf("got %v, expected %v", ttAns, tt.out)
+			}
+		})
+	}
+
+}
+
+// TestUstimeCast tests if casting was correct
+func TestUstimeCast(t *testing.T) {
+	t.Parallel()
+
+	var num uint64 = 11471851221
+	var expected uint32 = 2881916629
+	if uint32(num) != expected {
+		t.Errorf("expected %v, got : %v", expected, uint32(num))
+	}
+}
+
+// TestValidInitialPacket will send a packet via UDP, and check if it matches
+// The size and the Magic number field that needs to be equal.
+// This mocks the initial packet sent in Isoping.
+func TestValidInitialPacket(t *testing.T) {
+	client := NewInstance()
+	client.Start([]string{":4948"})
```

Comment:
Binding to a specific port is likely to fail on a busy CI system which is running lots of tests in parallel. Most tests bind to port 0, which means it will pick a random ephemeral port and every once in a while one of them will pick port 4948.

If you add a parameter to initServer for the port number, so you can pass in 4948 for production and 0 for tests, then:
- fetch the port number it actually used with srv.Conn.Addr().(UDPAddr).Port
- pass that port number to client.Start()

Then the test will be robust to having port 4948 be unavailable.